elements/pf-select/pf-select.ts

```diff
@@ -0,0 +1,517 @@
+      event.stopImmediatePropagation();
+      this.close();
+    } else if (this._input) {
+      this._input.value = '';
+
+      // set focus on input when a new chip is added
+      if (this.hasChips) {
+        this._input?.focus();
+      }
+    }
+  }
+
+  /**
+   * handles listbox option being created and creates a new "create option"
+   */
+  #onOptionCreated() {
+    this.#addCreateOption();
+    this.filter = '';
+    this.#updateCreateOptionValue();
+  }
+
+  /**
+   * handles typeahead combobox input event
+   */
+  #onTypeaheadInput() {
+    // update filter
+    this.#updateCreateOptionValue(this._input?.value || '');
+    if (this._listbox && this.filter !== this._input?.value) {
+      this.filter = this._input?.value || '';
+      if (this.#toggle) {
+        this.#toggle.open();
+      }
+    }
+  }
+
+  /**
+   * updates create option text to match input value
+   */
+  #updateCreateOptionText() {
+    const optionText = `${this.#createOption.getAttribute('value') || ''}`;
+    this.#createOption.innerHTML = optionText;
+    const [optionExists] = (this.options || []).filter(option => {
+      return option !== this.#createOption &&
+        (optionText === option?.textContent || this.#createOption.getAttribute('value') === option?.getAttribute('value'));
+    });
+    const createOptionText = !this.typeahead || optionText === '' || !!optionExists ? '' : this.createOptionText;
+    this.#createOption.createOptionText = createOptionText;
+  }
+
+  /**
+   * updates create option value to match current filter
+   */
+  #updateCreateOptionValue(inputValue: string = this.filter) {
+    let optionValue = inputValue || '';
+    if (optionValue === '*') {
+      optionValue = '';
+    }
+    this.#createOption.setAttribute('value', optionValue);
+    this.#updateCreateOptionText();
+  }
+
+  /**
+   * updates text indicating current value(s)
+   */
+  async #updateValueText() {
+    this.requestUpdate();
+    await this.updateComplete;
+
+    // reset input if chip has been added
+    if (this.hasChips && this._input?.value) {
+      const chip = this.shadowRoot?.querySelector(`pf-chip#chip-${this._input?.value}`) as HTMLElement;
+      if (chip && this._chipGroup) {
+        this._chipGroup.focusOnChip(chip);
+        this._input.value = '';
+      } else {
+        this._input.focus();
+      }
+    }
+  }
+
+  /**
+   * opens the dropdown
+   */
+  async open() {
+    await this.#toggle?.open(true);
+  }
+
+  /**
+   * sets focus
+   */
+  focus() {
+    const element = this._input && !this._input.hidden ? this._input
+      : this._toggle && !this._toggle.hidden ? this._toggle
+      : this._listbox;
+    element?.focus();
+  }
+
+  /**
+   * allows new options to be inserted
+   * @param option option to be inserted
+   * @param insertBefore optional: reference option before which new will be inserted; if blank new option inserted at end of list
+   */
+  insertOption(option: PfSelectOption, insertBefore?: PfSelectOption) {
```

Comment:
not sure i like the imperative api here.. would be better to just build from `set element.items = [...]`